src/Plugin/Validators/ValidDataFile.php

```diff
@@ -195,10 +260,160 @@ public function validateFile(int|null $fid) {
+   *       mime type are not compatible.
+   *
+   * @throws \Exception
+   *   - If the $validation_status parameter was not formatted properly.
+   *   - If the case string returned by the validator implied validation passed.
+   *   - If the case string returned by the validator is not recognized.
+   */
+  public static function processItemWithSimpleList(array $validation_status, array $tokens = []) {
+
+    // Check the format of the validation_status parameter.
+    ImportValidationHelper::checkValidationStatusArray($validation_status, 'ValidDataFile');
+
+    // Grab the default messages for all of our tokens (ones with default-msg).
+    $default_tokens = array_column(self::$mapping, 'default-msg', 'token');
+    // Combine our provided and our default token arrays. Because array_merge
+    // will overwrite values in the first array with values from the second
+    // array for the same keys, we provide our default tokens first.
+    $combined_tokens = array_merge($default_tokens, $tokens);
+
+    $container = \Drupal::getContainer();
+
+    // Get the current user in case we trigger a case that needs to log a
+    // message to the administrator.
+    $username = $container->get('current_user')->getAccountName();
+    $service_TripalLogger = $container->get('tripal.logger');
+
+    // Define our items array.
+    $items = [];
+
+    if (($validation_status['case'] == self::$mapping['case-invalid-fid']['dev-case']) ||
+        ($validation_status['case'] == self::$mapping['case-failed-fid']['dev-case'])) {
+
+      $message = $combined_tokens['case-invalid-fid'];
+      // Get the fid of the uploaded file.
+      $fid = $validation_status['failedItems']['fid'];
+
+      // Log a message for the administrator to help with debugging the issue.
+      $service_TripalLogger
+        ->info("The user $username uploaded a file with FID $fid using the Traits Importer, but could not import it as something is wrong with the filename/FID. More specifically, the case message '" . $validation_status['case'] . "' was reported.");
+    }
+    elseif ($validation_status['case'] == self::$mapping['case-empty-file']['dev-case']) {
+      $message = $combined_tokens['case-empty-file'];
+      $items = [
+        'Filename: ' . $validation_status['failedItems']['filename'],
+      ];
+    }
+    elseif (($validation_status['case'] == self::$mapping['case-unsupported-mime']['dev-case']) ||
+            ($validation_status['case'] == self::$mapping['case-unsupported-extension']['dev-case'])) {
+
+      $message = $combined_tokens['case-unsupported-mime'];
+      // Give more info to the user AND log a message to the administrator using
+      // these failed items:
+      $file_mime = $validation_status['failedItems']['mime'];
+      $file_extension = $validation_status['failedItems']['extension'];
+      $items = [
+        "The file extension indicates the file is \"$file_extension\" but our system detected the file is of type \"$file_mime\"",
+      ];
```

Comment:
This is a good contender for tokens within the items as brought up for discussion in https://github.com/TripalCultivate/TripalCultivate/pull/69#issuecomment-3037186270. For example, something like the token

`'explain-unsupported-mime' => 'The file extension indicates the file is [file_extension] but our system detected the file is of type [file_mime]',`

Then this is used like this

```suggestion
      $combined_tokens['file_mime'] = $validation_status['failedItems']['mime'];
      $combined_tokens['file_extension'] = $validation_status['failedItems']['extension'];
      $items = [
        $combined_tokens['explain-unsupported-mime'],
      ];
```

Reply:
Implemented the token in commit https://github.com/TripalCultivate/TripalCultivate/pull/69/commits/d51fc9afdb9478831707b62c710d1b7999c52072